pkg/asset/agent/image/agentimage.go

```diff
@@ -103,29 +108,75 @@ func (a *AgentImage) updateIgnitionImg(ignition []byte) error {
 		return err
 	}
 
-	ignitionImgPath := filepath.Join(a.tmpPath, "images", "ignition.img")
-	fi, err := os.Stat(ignitionImgPath)
+	// Defining a struct for the ignition info
+	type IgnInfo struct {
+		File   string `json:"file"`
+		Length int64  `json:"length"`
+		Offset int64  `json:"offset"`
+	}
+	var ignInfo IgnInfo
+
+	// Reading the ignition details from igninfo.json file into the struct
+	ignInfoJSONPath := filepath.Join(a.tmpPath, "coreos", "igninfo.json")
+	ignInfoJSONData, err := os.ReadFile(ignInfoJSONPath)
 	if err != nil {
+		logrus.Warnf("Failed to read json %s", ignInfoJSONPath)
+		return err
+	}
+	if err := json.Unmarshal(ignInfoJSONData, &ignInfo); err != nil {
+		logrus.Warnf("Failed to umarshal json: %s", err)
 		return err
 	}
 
-	// Verify that the current compressed ignition archive does not exceed the
-	// embed area (usually 256 Kb)
-	if len(ignitionBuff) > int(fi.Size()) {
-		return fmt.Errorf("ignition content length (%d) exceeds embed area size (%d)", len(ignitionBuff), fi.Size())
+	// Fetching the ignition details from the struct
+	ignImagePath := filepath.Join(a.tmpPath, ignInfo.File)
+	ignStartOffset := ignInfo.Offset
+	ignMaxLength := ignInfo.Length
+
+	// Verify that the compressed ignition buffer archive does not exceed the embed area (usually 256 Kb)
+	if ignMaxLength == 0 {
+		// Checking the file size if no ignition length is provided
+		fileInfo, err := os.Stat(ignImagePath)
+		if err != nil {
+			return err
+		}
+
+		if len(ignitionBuff) > int(fileInfo.Size()) {
+			return fmt.Errorf("ignition content length (%d) exceeds embed area size (%d)", len(ignitionBuff), fileInfo.Size())
+		}
+	} else if int64(len(ignitionBuff)) > ignMaxLength {
+		return fmt.Errorf("Ignition content length (%d) exceeds embed area size (%d)", len(ignitionBuff), ignMaxLength)
 	}
 
-	ignitionImg, err := os.OpenFile(ignitionImgPath, os.O_WRONLY, 0)
+	// Open the image file with permissions to seek and write the ignition content
+	ignitionImg, err := os.OpenFile(ignImagePath, os.O_RDWR, 0644)
 	if err != nil {
 		return err
 	}
 	defer ignitionImg.Close()
 
+	// Adjusting the ignition image file to the offset of ignition content (in case of cdboot.img)
+	_, err = ignitionImg.Seek(ignStartOffset, 0)
```

Comment:
As mentioned above, is there a case where the offset is different from 0 (not sure if it's a peculiarity of the s390x arch version)?

Reply:
@andfasano ,yeah for s390x the offset length differs. Below is THE reference on how igninfo.json looks for s390x

For reference the ignition info file looks like this on s390x:

{ "file": "images/cdboot.img", "length": 262144, "offset": 66086252 }